[28956] Resonsive forms

ingamai opened the pull request:
All forms should be responsive which means that they have, at best, a maximum width of the screen size and should be able to break into new lines, when the space is needed, without any breakpoints and therefore adapt themselves to every screen size. 

### **To-Do**
- [x] Make forms break automatically in new lines, when they need more space (then we will be able to remove a lot of mobile extra styles)
- [x] Adapt mobile: Forms should mostly break into a new line (except of checkboxes)
- [x] Forms should all break into a new line at the same time
- [x] Place the form annotations (has to work before and after flex wrap)

- [x] Make all **filters** responsive

- [x] **Cost reports** forms
- [x] Remove inline style width of cost reports

- [x] **Wiki History** scrollbar

- [x] Remove fixed width of Multiselects
- [ ] Position close buttons central

- [ ] Check forms with .grid-block classes

- [ ] Multiple fields per row -> form-grouping and affixes margin

- [ ] Check browser

https://community.openproject.com/projects/openproject/work_packages/28956
https://community.openproject.com/projects/openproject/work_packages/29033

ingamai committed:
Resonsive forms for every screen size

ingamai committed:
Administration forms adapted

ingamai committed:
Merge branch 'dev' into fix/unresponsive-forms

ingamai committed:
Merge branch 'dev' into fix/unresponsive-forms

ingamai committed:
Advanced filters responsive with grid layout

ingamai committed:
Work package filter width and page consistent with other sites

ingamai committed:
Small changes: custom actionsbuttons, my page form, admin plugins

ingamai committed:
Advanced and simple filters

ingamai committed:
Cost reports filter width animation removed and height/scrollbars fixed

ingamai committed:
Forms sass file reduced

ingamai committed:
Various small bugfixes

HDinger commented on app/assets/stylesheets/layout/work_packages/_table.sass:
```diff
@@ -54,29 +50,23 @@
   .work-packages--filters-optional-container
-    // not flex-item
-    height: auto
+    max-height: 40vh
```
This imitation leads pretty fast to a scrollbar within the filter. I think it would be better if the filter expand to their size and let either the table (like before) or the whole page scroll.

<img width="1429" alt="bildschirmfoto 2018-11-22 um 08 50 34" src="https://user-images.githubusercontent.com/7457313/48888781-0c7c8f80-ee34-11e8-90ac-96f8668cbad1.png">

ingamai replied:
I reverted my changes of this. There was a max-width of 50vh set before.

HDinger commented on app/assets/stylesheets/layout/work_packages/_mobile.sass:
```diff
@@ -132,6 +128,9 @@
       font-size: 1rem
 
 
+      .work-package--new-state
```
This rule is never applied. Perhaps it can be removed?

HDinger commented on app/assets/stylesheets/layout/_grid.sass:
```diff
@@ -37,3 +37,12 @@
   &.-visible-overflow
     overflow: visible
+
+.tab-content
```
Please use the foundation classes which do exactly that for you. So in the html add `large-6 small-12` to the `grid-content` class.

HDinger commented on app/assets/stylesheets/content/_simple_filters.sass:
```diff
@@ -68,100 +68,14 @@ $filters--border-color: $gray !default
-    .button
-      margin-bottom: 0rem
-
-.simple-filters--filters.-columns-3
```
Please remove the `-columns-3` class also from the LSG and the HTML.

HDinger commented on app/assets/stylesheets/content/_simple_filters.sass:
```diff
@@ -68,100 +68,14 @@ $filters--border-color: $gray !default
-    .button
-      margin-bottom: 0rem
-
-.simple-filters--filters.-columns-4
```
Same here. Please remove all code appearances of this class.

ingamai replied:
I hope I found all appearances in the project. Also, I updated the help texts in the .lsg files.

HDinger commented on app/assets/stylesheets/content/_forms.sass:
```diff
@@ -548,6 +480,9 @@ fieldset.form--fieldset
     .select2-container
       max-width: initial
 
+      &.colors-autocomplete
```
I did not know we have a autocompleter for colors 😳 Where is it used?

ingamai replied:
It's used for the color selection, e.g. for a work package type
![screen shot 2018-11-14 at 16 35 15](https://user-images.githubusercontent.com/39332861/49151961-a69b7680-f311-11e8-8c4d-775ca1042754.png)

HDinger commented on app/assets/stylesheets/content/_forms.sass:
```diff
@@ -430,59 +402,20 @@ fieldset.form--fieldset
 .form--field-instructions
   @extend %form--field-after-container
   font-style: italic
-  max-width: 500px // improve readability
```
I know that you removed this because the text is otherwise shown behind the input field and not below. However the way it looks now is also  not really good. Could you please check for a way to limit the width again?
<img width="1187" alt="bildschirmfoto 2018-11-22 um 09 55 33" src="https://user-images.githubusercontent.com/7457313/48891925-0b9c2b80-ee3d-11e8-8a7f-1bff5a8a6f35.png">

ingamai replied:
The instructions are now placed directly under the form fields and have a limited width again.

HDinger commented on app/assets/stylesheets/content/_forms.sass:
```diff
@@ -367,30 +362,13 @@ fieldset.form--fieldset
-  white-space: normal
-  overflow-y: visible
-  padding:      0 1rem 0 0
+  width: 200px
```
You know how I think about fixed widths. Sooner or later this will com back to us. However I did not find any issues because of this, yet. So it is fine for me.

ingamai replied:
Due to the grid layout, I could remove this now.

HDinger commented on app/assets/stylesheets/content/_advanced_filters.sass:
```diff
@@ -29,88 +29,61 @@
-
-.advanced-filters--filters
+.advanced-filters--filters,
+.simple-filters--filters
```
The file is named `advanced-filters`, so please do not mix the styles here. This will only make it harder in the future. If they share the same style let the advanced filters inherit the styles from the simple filters.

HDinger commented on app/assets/stylesheets/content/_advanced_filters.sass:
```diff
@@ -29,88 +29,61 @@
 // The type="text" is required to be more specific
 .advanced-filters--text-field[type="text"],
-.advanced-filters--date-field[type="text"]
+.advanced-filters--date-field[type="text"],
```
Please remove the comma.

HDinger reviewed (changes requested):
Hey @ingamai 

Good work. 👍  I think we may be able to remove a bit more of old code. (Have a look at the other comments) There are a few issues I found:

* The text inputs should not span the whole width. This looks strange.
<img width="1379" alt="bildschirmfoto 2018-11-22 um 08 37 43" src="https://user-images.githubusercontent.com/7457313/48893366-7a2eb880-ee40-11e8-81f8-97c0e261244c.png">

* On the 2FA screen the input is a bit too small
<img width="555" alt="bildschirmfoto 2018-11-22 um 09 23 54" src="https://user-images.githubusercontent.com/7457313/48893429-9b8fa480-ee40-11e8-917f-215f9587fd20.png">

* The simple filters do not span the whole width on mobile. 
<img width="520" alt="bildschirmfoto 2018-11-22 um 09 31 55" src="https://user-images.githubusercontent.com/7457313/48893505-d09bf700-ee40-11e8-9149-d1bacb11b401.png">

* On desktop we should use the given width and show multiple filters per row as it was before. 
<img width="1115" alt="bildschirmfoto 2018-11-22 um 10 25 36" src="https://user-images.githubusercontent.com/7457313/48893616-0ccf5780-ee41-11e8-9014-95d0094faa33.png">


* The filters still have a problem when the selector can be quite long. For example when there is a UserSelect and a User with a long name.
<img width="556" alt="bildschirmfoto 2018-11-22 um 09 33 31" src="https://user-images.githubusercontent.com/7457313/48893723-4e600280-ee41-11e8-87a6-2ca78d3cee14.png">

* The "Add filter" label breaks quite ugly. Please have a look whether we shall avoid a line break here, give more space, or something else. When doing so please check whether we can remove this code here: https://github.com/opf/openproject/blob/dev/app/assets/stylesheets/content/_simple_filters.sass#L55
<img width="222" alt="bildschirmfoto 2018-11-22 um 09 35 40" src="https://user-images.githubusercontent.com/7457313/48893873-a1d25080-ee41-11e8-9ca7-f5c77ec63657.png">

* In my opinion the input fields should span the whole width on mobile. Otherwise there are really small. 
<img width="527" alt="bildschirmfoto 2018-11-22 um 09 37 02" src="https://user-images.githubusercontent.com/7457313/48893945-cd553b00-ee41-11e8-995a-04976125f472.png">

* The inline subject form is still to small:
<img width="1207" alt="bildschirmfoto 2018-11-22 um 14 32 55" src="https://user-images.githubusercontent.com/7457313/48906102-bf181680-ee63-11e8-9f78-b54c7831434e.png">


* Finally the forms should break all together, otherwise the layout is messy.
<img width="501" alt="bildschirmfoto 2018-11-22 um 10 34 24" src="https://user-images.githubusercontent.com/7457313/48894119-33da5900-ee42-11e8-90d8-7f86c7f0cf40.png">

I did not have a look at the browsers yet, I will do this once you've included the changes.

ingamai committed:
Requested changes: e.g. forms as grid, multiple filters per row

ingamai committed:
'Add Member' form fixed

ingamai committed:
Mobile breakpoint for advanced and simple filters

ingamai committed:
Vertical layout fixed (e.g. Login & Sign in); small review changes)

ingamai commented:
Cause it was quite a lot, here are all changes I made until the review:

- Advanced filters grid (labels not breaking anymore -> but still a problem with very long labels)
- Input fields shorter
- Simple filters: multiple fields per row (also grid layout)
- Forms as grid, so that they are wrapping at the same time (which is unfortunately causing this bug https://community.openproject.com/projects/openproject/work_packages/29106, for which we don’t have an easy and quick solution yet)
- Viewport max-height reverted to 50vh
- Form field instructions positioned under form fields again (with limited width)
- Members form for ‘Add member’ fixed
- Login and Sign-in form fields fixed
- Mobile styles for advanced and simple filters
- Outdated code removed, .lsg file updated

- Fixed width on form labels is unfortunately still there, but I changed it to a max-width and set it from 200px to 250px. We could change that as soon as we have media queries on elements.

ingamai commented:
Problems I noticed going through the style guide:
- General difficulties with grid-block and grid-content (maybe this has to be checked more precisely again)
-  Multiple fields per row (form—grouping works with flex and affixes margin missing)

Generally speaking, the grid layout is good to give them all the same structure and behaviour. But unfortunately we have a lot of exceptions, where it works different and doesn't have the "normal" form structure.

HDinger committed:
Merge branch 'dev' into fix/unresponsive-forms

HDinger commented:
The complexity is too high to be sure that this does not break somewhere. I therefore created a new PR which handles the bugfixes and the refactoring of the filters. I let the forms untouched there. https://github.com/opf/openproject/pull/6876

I will close this PR in favour of the new one. In case we want to address the forms later we can come back to this code.
